### ISSUE TITLE
waiting for completion of hook and hook never succeds

### ISSUE THREAD

**Issue by rajivml** [argoproj/argo-cd#6880]
HI,

We are seeing this issue quite often where app sync is getting stuck in "waiting for completion of hook" and these hooks are never getting completed

As you can see the below application got stuck on secret creation phase and some how that secret never got created

![image](https://user-images.githubusercontent.com/46923567/127797111-89cb30ef-3c97-447b-9f60-ad3bafa6bbd0.png)

Stripped out all un-necessary details. Now this is how the secret is created and used by the job.

```
apiVersion: v1
kind: Secret
metadata:
  name: {{ include "xxx.fullname" . }}-migrations-{{ .Chart.AppVersion }}
  annotations:
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
    helm.sh/hook-weight: "-5"
type: Opaque
data:
  xxxx

apiVersion: batch/v1
kind: Job
  annotations:
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation
    helm.sh/hook-weight: "-4"
spec:
      annotations:
        sidecar.istio.io/inject: "false"
    spec:
        - name: app-settings
          configMap:
            name: {{ include "xxx.fullname" . }}-migrations-{{ .Chart.AppVersion }}
        - name: app-secrets
          secret:
            secretName: {{ include "xxx.fullname" . }}-migrations-{{ .Chart.AppVersion }}
```

kubectl -n argocd logs argocd-server-768f46f469-j98h6 | grep xxx-migrations   - No matching logs
kubectl -n argocd logs argocd-repo-server-57bdbf899c-9lxhr | grep xxx-migrations - No matching logs
kubectl -n argocd logs argocd-repo-server-57bdbf899c-7xvs7 | grep xxx-migrations - No matching logs
kubectl -n argocd logs argocd-server-768f46f469-tqp8p | grep xxx-migrations - No matching logs

[testadmin@server0 ~]$ kubectl -n argocd logs argocd-application-controller-0 | grep orchestrator-migrations
time="2021-08-02T02:16:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:16:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:19:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:19:26Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:22:17Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:22:17Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:22:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:25:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:25:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:28:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:28:26Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:31:25Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx
time="2021-08-02T02:31:26Z" level=info msg="Resuming in-progress operation. phase: Running, message: waiting for completion of hook /Secret/xxx-migrations-0.0.19-private4.1784494" application=xxx

Environment:
* 3 Node RKE2 Cluster
* OS: RHEL 8.4
* K8's setup on Azure VM's

ArgoCD Version: 2.0.1

Please let me know in case of any other info required

**Comment by rajivml** [comment]
I have terminated the app sync and re-synced it again and the sync is successful now but this can't happen because if it happens CI / CD runs and also the automation that we have done to install apps via argoCD CLI would fail.

**Comment by alexmt** [comment]
I suspect this is fixed by https://github.com/argoproj/argo-cd/pull/6294 . The fix is available in https://github.com/argoproj/argo-cd/releases/tag/v2.0.3 . Can you try upgrading please?

**Comment by rajivml** [comment]
sure, thanks, we recently upgraded our develop to to use 2.0.5 and this happened on our prod build which is on 2.0.1. I will see if this can repro on our dev branch. Thanks !

**Comment by om3171991** [comment]
@alexmt - We are using the below version of ArgoCD and seeing the same issue with Contour helm. Application is waiting for  PreSync Job to complete whereas on a cluster I can see the job is completed.

{
    "Version": "v2.1.3+d855831",
    "BuildDate": "2021-09-29T21:51:21Z",
    "GitCommit": "d855831540e51d8a90b1006d2eb9f49ab1b088af",
    "GitTreeState": "clean",
    "GoVersion": "go1.16.5",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KsonnetVersion": "v0.13.1",
    "KustomizeVersion": "v4.2.0 2021-06-30T22:49:26Z",
    "HelmVersion": "v3.6.0+g7f2df64",
    "KubectlVersion": "v0.21.0",
    "JsonnetVersion": "v0.17.0"
}

**Comment by illagrenan** [comment]
I have the same problem in version `2.2.0`.

**Comment by pseymournutanix** [comment]
I have the same problem on the 2.3.0 RC1 as well

**Comment by jaydipdave** [comment]
The PreSync hook, PostSync hook, and "Syncing" (while No Operation Running) are the only long pending major issues in ArgoCD at the moment.

**Comment by aslamkhan-dremio** [comment]
Hello. I am still seeing this in v2.2.4. PreSync hook is scheduled, Job starts, runs to completion, Argo sits there spinning "Progressing" until terminated. To work around it, we are terminating the op and using 'sync --strategy=apply' (disabling the hook) and running our job out of band.

Kube events during the sync confirm the job success. I no longer see the job/pod (per those events) if I check the namespace directly.

`LAST SEEN   TYPE     REASON                   OBJECT                                                      MESSAGE
22m         Normal   Scheduled                pod/dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l   Successfully assigned dcs-prodemea-ns/dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l to gke-service-nap-e2-standard-8-1oj503q-5bf9adda-f9t6
22m         Normal   Pulling                  pod/dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l   Pulling image "gcr.io/dremio-1093/accept-release:v3"
21m         Normal   Pulled                   pod/dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l   Successfully pulled image "gcr.io/dremio-1093/accept-release:v3" in 48.040095979s
21m         Normal   Created                  pod/dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l   Created container dcs-artifact
21m         Normal   Started                  pod/dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l   Started container dcs-artifact
22m         Normal   SuccessfulCreate         job/dcs-artifact-promoter0ba458b-presync-1645132298         Created pod: dcs-artifact-promoter0ba458b-presync-1645132298-hqm9l`

![image](https://user-images.githubusercontent.com/79135023/154610560-28c0b01b-fc39-4520-b94d-0946887efc66.png)

Let me know if I can provide any diagnostics to help.

**Comment by MariaJohny** [comment]
We face the same issue in 2.2.5 as well.

**Comment by MariaJohny** [comment]
> I suspect this is fixed by #6294 . The fix is available in https://github.com/argoproj/argo-cd/releases/tag/v2.0.3 . Can you try upgrading please?

Does it work with 2.0.3 or 2.2.2?

**Comment by ceguimaraes** [comment]
I can confirm the error was fixed on 2.0.3. We recently upgraded to 2.3.3 and we are experiencing the error again.

**Comment by warmfusion** [comment]
We're seeing a similar issue on the syncfailed hook which means we can't actually terminate the sync action.

The job doesn't exist in the target namespace, and we've tried to trick argo by creating a job with the same name, namespace, and annotations as we'd expect to see with a simple echo "done' action but nothing is helping.

![image](https://user-images.githubusercontent.com/2294568/174116102-c043b49c-7064-4cec-8d0b-998cb0d2881d.png)

**Comment by margueritepd** [comment]
To add some information here, we are running into the same issue ("waiting for completion of hook" when the hook has already completed), and it happens when we are attempting to sync to a revision that is not the targetRevision for the app. When we sync an app with hooks to the same revision as the targetRevision, we do not run into this.

<details><summary>expand for argo version</summary>

```
{
    "Version": "v2.3.4+ac8b7df",
    "BuildDate": "2022-05-18T11:41:37Z",
    "GitCommit": "ac8b7df9467ffcc0920b826c62c4b603a7bfed24",
    "GitTreeState": "clean",
    "GoVersion": "go1.17.10",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KsonnetVersion": "v0.13.1",
    "KustomizeVersion": "v4.4.1 2021-11-11T23:36:27Z",
    "HelmVersion": "v3.8.0+gd141386",
    "KubectlVersion": "v0.23.1",
    "JsonnetVersion": "v0.18.0"
}
```
</details>

We are running 2 application-controller replicas in HA setup as per https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/ . I have verified we do not have a leftover pre-stateful-set instance also running.

**Comment by lacarvalho91** [comment]
i had this problem when i was configuring resource inclusions, i wrote down what happened here: https://github.com/argoproj/argo-cd/issues/10756#issuecomment-1265488657

**Comment by cscorley** [comment]
We resolved this symptom on v2.4.12+41f54aa for Apps that had many Pods by adding a resource exclusion along these lines to our config map:

```
data:
  resource.exclusions: |
    - apiGroups:
        - '*'
      kinds:
        - 'Pod'
      clusters:
        - '*'
```

Prior to this, we would have pre-sync job hooks never completing in the ArgoCD UI, but would have actually be completed in Kubernetes.  Sometimes, invalidating the cluster cache would help Argo recognize the job was completed, but most of the time not.

We believe the timeouts were related to needing to enumerate an excessive amount of entities and just simply never could finish before the next status refresh occurred.  We do not utilize viewing the status of Pods through ArgoCD UI, so this solution is fine for us.

**Comment by DasJayYa** [comment]
We had this issue and it was relating to a customers Job failing to initialise due to a bad secret mounting. You can validate this by checking the events in the namespace the job is being spun up to see if its failing to create.

**Comment by dejanzele** [comment]
Hello Argo community :)

I am fairly familiar with ArgoCD codebase and API, and I'd happily try to repay you for building such an awesome project by trying to have a stab at this issue, if there are no objections?

**Comment by williamcodes** [comment]
I would also highly appreciate that!

**Comment by vumdao** [comment]
I'm seeing this issue with `v2.6.1+3f143c9`

**Comment by linuxbsdfreak** [comment]
I am also seeing this issue when installing kubevela with argocd with the version **v2.6.1+3f143c9**

```
    message: >-
      waiting for completion of hook
      /ServiceAccount/kube-vela-vela-core-admission and 3 more hooks
```

**Comment by micke** [comment]
We also had this issue and it was resolved once we set `ARGOCD_CONTROLLER_REPLICAS`.

Instructions here: https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller

> If the controller is managing too many clusters and uses too much memory then you can shard clusters across multiple controller replicas. To enable sharding increase the number of replicas in argocd-application-controller StatefulSet and repeat number of replicas in ARGOCD_CONTROLLER_REPLICAS environment variable. The strategic merge patch below demonstrates changes required to configure two controller replicas.

**Comment by boedy** [comment]
I rolled back from `1.6.1` to `1.5.10`. Both versions keep waiting for completion of hook, which has already successfully completed.

I also tried @micke's recommendation (changing the `ARGOCD_CONTROLLER_REPLICAS` from 1 to 3). Doesn't make a difference unfortunately.

**Comment by boedy** [comment]
I just figured out what was causing Argo to freeze on the hook. In my case the specific hook had `ttlSecondsAfterFinished: 0` defined in the spec. Through Kustomize I removed this field:

```yaml
patches:
  - target:
      name: pre-hook
      kind: Job
    path: patches/hook.yaml
```
```yaml
# patches/hook.yaml
- path: "/spec/ttlSecondsAfterFinished"
  op: remove
```

Afterwards the chart finally went through! It's still a bug that should be addressed, I'm just sharing this for others to work around it.

**Comment by zfrhv** [comment]
I had this problem when i had CR which CRD is still not created, and a job with Sync hook

so argocd couldnt apply the custom resource because there was no crd yet, and the hook started and then disappeared. I guess because argo retries to sync the CR and it also restarts the hook somehow
(btw I was using SkiDryRunOnMissingResource for the CR)

so I just did that the hook will be PostSync.
the CR was retrying until the crd were created, and only after the CR was successfully created then the PostSync hook started and completed successfully

**Comment by adlnc** [comment]
Encountered similar behavior as described in this issue while upgrading from v2.5.12+9cd67b1 to **v2.6.3+e05298b**
pre-upgrade hooks on different applications with various number of pods and jobs had same symptoms.
Sync operation is running forever. 
Had the feeling this random event seems to appear more frequently while using the argo cli.

**Comment by chaehni** [comment]
Observing the same issue on v2.6.2. Post-Sync hook never completes even though the corresponding pod exited successfully.

**Comment by 0Bu** [comment]
Same with 2.6.5

**Comment by Tarasovych** [comment]
Same issue with v2.7.3+e7891b8.dirty

**Comment by asaf400** [comment]
Same issue in v1.3.3.7 and also in version v6.9..
This issue was opened on Aug 2, 2021, we are now at 2023, please bump this comment via emoji so I can see it in my inbox in 2042

**Comment by Whisper40** [comment]
Hello, same here version 2.5.0.
A controller restart has been needed to be able to sync again.

**Comment by javydekoning** [comment]
Facing the same on `v2.7.9` while (re)deploying https://artifacthub.io/packages/helm/prometheus-community/kube-prometheus-stack

**Comment by asaf400** [comment]
> Looking at what appears to be the same issue on ArgoCD v2.6.7. I've killed all controller pods, server and repo server at the same time, hoping that ArgoCD would start behaving. But to no avail. I believe the reason for it to start behaving like this is in the first place was a `ImagePullBackOff` on the `Job` image.

We had to completely exclude all jobs from argo cd via resource exclusion global config:
https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#resource-exclusioninclusion

And we migrated all jobs on repo to cronjobs with `suspend: true`
BUT fair warninig, Due to a k8s bug sometimes cronjobs may be triggered when changing spec, INCLUDING changing `suspend: false` to `suspend: true` - yes it's stupid like that..

I think it's this one, but there are others as well..
https://github.com/kubernetes/kubernetes/issues/63371

**Comment by turesheim** [comment]
Ended up executing `kubectl delete application <app>` after fixing the reason for the `ImagePullBackOff`. This resulted in the stuck job disappearing and the application resynced and in order. So there is a decent way get rid of the stuck `Job`.

**Comment by jDmacD** [comment]
> I just figured out what was causing Argo to freeze on the hook. In my case the specific hook had `ttlSecondsAfterFinished: 0` defined in the spec. Through Kustomize I removed this field:
> 
> ```yaml
> patches:
>   - target:
>       name: pre-hook
>       kind: Job
>     path: patches/hook.yaml
> ```
> 
> ```yaml
> # patches/hook.yaml
> - path: "/spec/ttlSecondsAfterFinished"
>   op: remove
> ```
> 
> Afterwards the chart finally went through! It's still a bug that should be addressed, I'm just sharing this for others to work around it.

@boedy You're a Saint. I've been staring at [envoyproxy/gateway](https://github.com/envoyproxy/gateway) for two weeks.

**Comment by MarkDPierce** [comment]
> Ended up executing `kubectl delete application <app>` after fixing the reason for the `ImagePullBackOff`. This resulted in the stuck job disappearing and the application resynced and in order. So there is a decent way get rid of the stuck `Job`.

We just had this issue and we did a similar thing where we deleted the namespace of the app. It was removed from k8s but ArgoCD still saw it.

Running `kubectl patch app APP_NAME -n argocd -p '{"metadata":{"finalizers":null}}' --type merge` or what ever namespace you deploy your argocd apps into will help the delete go through. Then you can deploy it again. You should be careful when taking this action on stateful apps though.

**Comment by KilJaeeun** [comment]
hello ! My preinstall job is still syncing 10 hours after it finished. I was told to delete the ttl strategy, so I deleted the ttl in the pre install job, but I'm still having the same problem. Does this mean I should change the ttl strategy on the argocd side? Is this something that hasn't been resolved yet in the latest update?


cc. @boedy  
> I just figured out what was causing Argo to freeze on the hook. In my case the specific hook had `ttlSecondsAfterFinished: 0` defined in the spec. Through Kustomize I removed this field:
> 
> ```yaml
> patches:
>   - target:
>       name: pre-hook
>       kind: Job
>     path: patches/hook.yaml
> ```
> 
> ```yaml
> # patches/hook.yaml
> - path: "/spec/ttlSecondsAfterFinished"
>   op: remove
> ```
> 
> Afterwards the chart finally went through! It's still a bug that should be addressed, I'm just sharing this for others to work around it.

**Comment by imroc** [comment]
Issue still exists in `argocd:v2.10.1`
![image](https://github.com/argoproj/argo-cd/assets/7448852/0fe32fe3-f521-4488-872e-18faacf1902c)

**Comment by Lp-Francois** [comment]
Observing the same issue with argocd `v2.9.2+c5ea5c4`

From a manifest like this job: https://github.com/aws-samples/eks-gitops-crossplane-argocd/blob/main/crossplane-complete/templates/2-aws-provider.yaml#L28

<img width="893" alt="Screenshot 2024-03-01 at 10 31 37" src="https://github.com/argoproj/argo-cd/assets/32224751/2ed35e6a-cb21-451f-82fd-ad8f147684c7">
<img width="681" alt="Screenshot 2024-03-01 at 10 31 45" src="https://github.com/argoproj/argo-cd/assets/32224751/a93097d0-4084-4164-8409-ae65a0192f8d">

**Comment by michalschott** [comment]
Still happens in `v2.10.2+fcf5d8c`

**Comment by asaf400** [comment]
> Still happens in `v2.10.2+fcf5d8c`

Haha 🤣 , Good one 👍

**Comment by huguesalary** [comment]
Running into this issue on `v2.10.1+a79e0ea` as well

**Comment by fabi-alstom** [comment]
I'm using v2.11.0+f287dab and hit the same problem with whatever version of kube-prometheus-stack from 45.0.0 on. Last testing was with the most recent one, 58.0.0, and still facing the issue. Sadly, no workaround worked for me yet

**Comment by JuniorJPDJ** [comment]
The same with envoy-gateway

**Comment by imroc** [comment]
> The same with envoy-gateway

Try:
```yaml
patches:
  - target:
      name: eg-gateway-helm-certgen
      kind: Job
    patch: |
      - path: "/spec/ttlSecondsAfterFinished"
        op: remove
  - target:
      name: envoy-gateway
      kind: Deployment
    patch: |
      - path: "/spec/template/spec/containers/0/resources"
        op: remove
```

**Comment by JuniorJPDJ** [comment]
@roc how am I supposed to use it with helm?

**Comment by imroc** [comment]
> ```yaml
> helm
> ```

You need kustomize, if you must use helm, you can combine helm with kustomize, use `helmCharts` in `kustomization.yaml`  to include envoygateway chart, and also add patches in `kustomization.yaml`  .

**Comment by JuniorJPDJ** [comment]
I figured it out - envoy gateway allows setting this parameter directly on helm values - no need for kustomize:
```
      certgen:
        job:
          ttlSecondsAfterFinished: ~
```
Or set it to `60` and it doesn't bug ArgoCD too.

**Comment by florentinchaussoy** [comment]
FYI - present in v2.10.7 but not in v2.10.4.

**Comment by joelmccoy** [comment]
FYI - Running into the same issue with longhorn on ArgoCD v2.10.1+a79e0ea.  
![image](https://github.com/argoproj/argo-cd/assets/14078350/07ae9ce7-3d55-423b-8b8c-1a587ff6a6bb)

**Comment by jkleinlercher** [comment]
> I'm using v2.11.0+f287dab and hit the same problem with whatever version of kube-prometheus-stack from 45.0.0 on. Last testing was with the most recent one, 58.0.0, and still facing the issue. Sadly, no workaround worked for me yet

Maybe a ttl > 0 is needed also in https://github.com/prometheus-community/helm-charts/blob/9c41858ac9714483638d78fb560577dc37e55875/charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/job-patch/job-createSecret.yaml#L19

just according to others who changed the ttl. Maybe someone can shed some light on this why changing the ttl helps in some charts?

**Comment by sambonbonne** [comment]
@jkleinlercher if I understand the issue, it seems that when the TTL is set to 0 ArgoCD does not have the time to detect the job succedeed and waits for it to finish indefinitely.

**Comment by jkleinlercher** [comment]
I also asked again in https://cloud-native.slack.com/archives/C01TSERG0KZ/p1714376880925509 about this issue. Would be happy if we could get some experts in to tell if "ttlSecondsAfterFinished: 0" in a hook job is a problem (or under which circumstances it could be a problem). Maybe then some off-the-shelf helm-charts could be reconfigured to solve this problem for everyone.

**Comment by jkleinlercher** [comment]
I no recognized that the 'ttlSecondsAfterFinished: 0' cannot be the root cause because this is not set in our situation. in kube-prometheus-stack there is a API condition around this setting which is not met in current clusters:

https://github.com/prometheus-community/helm-charts/blob/9c41858ac9714483638d78fb560577dc37e55875/charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/job-patch/job-createSecret.yaml#L17-L20

So there must be another root cause for the "stuck in pending deletion" situation.

**Comment by jkleinlercher** [comment]
okay locally on k3d cluster I can recreate the problem with kube-prometheus-stack but argocd didn't stuck in "pending deletion" but "running", although the job already finished and is already deleted and


![image](https://github.com/argoproj/argo-cd/assets/11465610/702a13e2-af61-48d3-a1b4-6d30462b2850)


```
kubectl get job -n monitoring sx-kube-prometheus-stack-admission-create
Error from server (NotFound): jobs.batch "sx-kube-prometheus-stack-admission-create" not found
```

**Comment by slashr** [comment]
Reporting the same problem on a k3s cluster trying to install the kube-prometheus-stack Helm chart latest version

<img width="1148" alt="image" src="https://github.com/argoproj/argo-cd/assets/2951124/277b6f88-b4fe-48f9-878d-191c64d59332">

**Comment by jkleinlercher** [comment]
Meanwhile I came across https://github.com/argoproj/argo-cd/issues/15292 and I wonder if this is the same problem. In

https://github.com/prometheus-community/helm-charts/blob/9c41858ac9714483638d78fb560577dc37e55875/charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/job-patch/job-createSecret.yaml#L9

is the same deletion policy as in the issue mentioned … sadly that the referenced PR https://github.com/argoproj/gitops-engine/pull/461 was never merged. @leoluz or @nazarewk is there some chance to help out on this?

**Comment by ptr1120** [comment]
Same problem now using Argo v2.11.0-rc3+20fd621 with current kube-prometheus-stack

**Comment by NAVRockClimber** [comment]
Same problem here 5 different clusters (K3S, Kubespray & Managed) on none I can deploy the current kube-prometheus-stack (Chart v58.3.1) using Argo-CD 2.10.9

**Comment by slashr** [comment]
Issue has been fixed

https://github.com/prometheus-community/helm-charts/pull/4510

**Comment by j809** [comment]
The job `job-createSecret.yaml` does not complete syncing in `58.3.2`. Setting `prometheusOperator.admissionWebhooks.patch.ttlSecondsAfterFinished` to 30s helped me solve the problem.

**Comment by prashant0085** [comment]
Still facing the issue when trying to deploy kube prom stack helm chart version `58.6.1` where `prometheusOperator.admissionWebhooks.patch.ttlSecondsAfterFinished` is set to `60`
Argocd version: `v2.9.0+9cf0c69`
![image](https://github.com/argoproj/argo-cd/assets/31355915/63c162c9-d9d9-4d62-9a94-84f3b6072d6b)
![image](https://github.com/argoproj/argo-cd/assets/31355915/8166a03b-dc69-4f06-a118-34a506f3d336)

**Comment by jsantosa-minsait** [comment]
> Still facing the issue when trying to deploy kube prom stack helm chart version `58.6.1` where `prometheusOperator.admissionWebhooks.patch.ttlSecondsAfterFinished` is set to `60` Argocd version: `v2.9.0+9cf0c69` ![image](https://private-user-images.githubusercontent.com/31355915/334807484-63c162c9-d9d9-4d62-9a94-84f3b6072d6b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwODgzMTcsIm5iZiI6MTcxNzA4ODAxNywicGF0aCI6Ii8zMTM1NTkxNS8zMzQ4MDc0ODQtNjNjMTYyYzktZDlkOS00ZDYyLTlhOTQtODRmM2I2MDcyZDZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMwVDE2NTMzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxMjBjMDNjYzc1NDI5MjM4MWJmYjAzMDgzNWRkMWZmYzQxYmQxMDA2MDE0MTYwZDNkNTVhM2I4Yjc0MDQ3YWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eIxCYk76blUYpXgLAst71PRHi0FdBNDeQwRprwEtXJQ) ![image](https://private-user-images.githubusercontent.com/31355915/334807628-8166a03b-dc69-4f06-a118-34a506f3d336.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwODgzMTcsIm5iZiI6MTcxNzA4ODAxNywicGF0aCI6Ii8zMTM1NTkxNS8zMzQ4MDc2MjgtODE2NmEwM2ItZGM2OS00ZjA2LWExMTgtMzRhNTA2ZjNkMzM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMwVDE2NTMzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlMzE1NTVkMDI4M2VlNTI2NGY4NGJiMjNhYjk3ZjYxNGM1MmE1NjAxNWU5NDZiN2ZlMWVlZDExNDg4YTEyOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3Uinwc1tKAkq1TBiHPqSiIMYblCWoqba1bFmwxY7Pjw)

I have the same issue, same configuration in ttlSecondsAfterFinished

**Comment by prashant0085** [comment]
@jsantosa-minsait Is by chance you have enabled istio sidecar injection ? that causes the pod to complete the patching and creation but the pod keeps on running.

**Comment by jsantosa-minsait** [comment]
> @jsantosa-minsait Is by chance you have enabled istio sidecar injection ? that causes the pod to complete the patching and creation but the pod keeps on running.

Hi @prashant0085 no, I don't.  I have Cilium installed and Kyverno with admissions controllers hooks that may alter or patch the resource. However this is not the case.

**Comment by ilabrovic** [comment]
Hi, experiencing this issue as well.
Environment: Openshift 4.14, Argocd v2.10.10+9b3d0c0, Postsynchook job.
Just a simple kustomization.yaml with 2 resources and a postsynchookjob
No Helm.
The job actually takes about 2 minutes to complete, but argocd only sets the job as finished after approx 10 minutes.
Tried different settings (0, 60, 120) on ttlSecondsAfterFinished in the job spec, but no change in behaviour

**Comment by azorahai3724** [comment]
> Hi, experiencing this issue as well. Environment: Openshift 4.14, Argocd v2.10.10+9b3d0c0, Postsynchook job. Just a simple kustomization.yaml with 2 resources and a postsynchookjob No Helm. The job actually takes about 2 minutes to complete, but argocd only sets the job as finished after approx 10 minutes. Tried different settings (0, 60, 120) on ttlSecondsAfterFinished in the job spec, but no change in behaviour Also monitored the memory and cpu usage of the argocd pods (controller, repo, applicationsetcontroller, etc), no pod even comes close to limitcpu or limitmemory, so no issue there...

We seem to experience a similar situation with the pre and post-sync hooked pods, argocd only sets them finished after circa 10 minutes even though they actually complete earlier.

**Comment by travis-jorge** [comment]
this happens for us at least once a day on our db jobs that are pre-sync hooks v2.11.5. Any suggestions on how to determine what is causing this?

**Comment by jkleinlercher** [comment]
Since ArgoCD helm chart now has also a webhook for redis we get this problem also with ArgoCD managing ArgoCD. It stays OutOfSync with this

waiting for completion of hook batch/Job/argocd-redis-secret-init


        - group: batch
          hookPhase: Running
          hookType: PreSync
          kind: Job
          message: job.batch/sx-argocd-redis-secret-init created
          name: sx-argocd-redis-secret-init
          namespace: argocd
          syncPhase: PreSync
          version: v1
```

that is really a problem …

**Comment by tico24** [comment]
Because I merged the PR which has magic words on it so GitHub closed this issue.

**Comment by mqxter** [comment]
still an issue. hoping for a fix.

**Comment by petrlebedev** [comment]
Also same issue with Karpenter helm chart 3.7.1: stucked on: `waiting for completion of hook batch/Job/karpenter-staging-post-install-hook`
Sad that i can't change chart)
<img width="933" alt="image" src="https://github.com/user-attachments/assets/e31f725a-408c-45df-afea-ac0afce5f3ad">

any workarounds you can share?

**Comment by tico24** [comment]
This is primarily due to Karpenter's (frankly rubbish) implementation. I have been battling errors in the 1.0.0 upgrade for the past two days. Warning: removing Argo from the mix doesn't make things much better. 

There as workarounds documented in the issues in the karpenter repo.

**Comment by petrlebedev** [comment]
@tico24 
`          kustomize:
            patches:
              - target:
                  group: batch
                  version: v1
                  kind: Job
                  name: karpenter-post-install-hook
                patch: |-
                  - op: replace
                    path: /spec/ttlSecondsAfterFinished
                    value: 30`
                    
                    https://github.com/aws/karpenter-provider-aws/issues/6765#issuecomment-2296557387

**Comment by RevealOscar** [comment]
Had the same issue although for me the service account was not being deleted with the same hooks. My job ttlSecondsAfterFinished was > 0. Unsure if this is the actual root issue but I added this annotations to my job/sa
```argocd.argoproj.io/sync-options: PruneLast=false ```
My reasoning being that the sync wave for pruning was going to be Last under the default. The chart needs the deletion to happen before any resources are upgraded. I have not noticed the issue on installation of charts. Per the argocd docs pruneLast would occur ```after the other resources have been deployed and become healthy``` so my assumption is argo holds on to the resource causing the sync to be delayed indefinitely. Been testing for a week without issue

**Comment by DoumLaberge** [comment]
I still have the issue first described in argocd v 2.14.2
![image](https://github.com/user-attachments/assets/3fcfac5f-5c90-4672-a8b9-33bdf9509010)

**Comment by jkleinlercher** [comment]
I have this problem in our pipeline where we create a kind cluster and manage ArgoCD via ArgoCD Helm chart almost every day (stuck in presync redis init job). If someone of Inuit, Redhat, Akuity,.. helps me how to debug/gather debug data I would be happy to to gather all data to troubleshoot this problem. @jessesuen @crenshaw-dev ? It seems solving this problem would help lots of people in the community and I would help as much as I can ..

**Comment by sandeep-ongit** [comment]
+1
same issue with v2.11.10+971c6d0

**Comment by juangb** [comment]
In my case, it was solved by deleting the sync option `- Replace=true` from the application, no idea why it was interfering with Argo, and hindering it to realize that the job was finished. The issue was with an ingress-nginx job with the following annotations:      
````yaml
---
apiVersion: batch/v1
kind: Job
metadata:
  name: ingress-nginx-admission-create
  namespace: ingress-nginx
  annotations:
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
spec:
  template:
    metadata:
      annotations:
        karpenter.sh/do-not-disrupt: 'true'
```

**Comment by kohbis** [comment]
Same issue with `v2.13.0+347f221`

**Comment by andrii-korotkov-verkada** [comment]
Would overriding `ttlSecondsAfterFinished` or Argo side to be > 0 if it's set to 0 solve the problem?

**Comment by kohbis** [comment]
> Would overriding ttlSecondsAfterFinished or Argo side to be > 0 if it's set to 0 solve the problem?
No, even if ttlSecondsAfterFinished is set to a value > 0, the Job remains in an incomplete state in Argo CD.

**Comment by dejanzele** [comment]
Hello awesome argo community,

I played around with this bug and implemented a proof-of-concept PR which proposes a fix for the stuck hook issue.

I kindly ask for feedback from the community and maintainers, as I would very much like to upstream it.

Issue - https://github.com/argoproj/argo-cd/issues/21055
PR - https://github.com/argoproj/gitops-engine/pull/646

**Comment by crenshaw-dev** [comment]
The solution @dejanzele talked about, and the one being worked on in https://github.com/argoproj/gitops-engine/pull/646, is to add a finalizer to all hook resources to prevent hook resources from being deleted before Argo CD can observe the final state. I'm curious what others think of this approach.

**Comment by dyatlov** [comment]
that aforementioned `10m` timeout seem to correlate with https://github.com/argoproj/argo-cd/blob/5b79c34c72300e6e2e6336051ce6992f6d54011c/controller/cache/cache.go#L89-L91

it appears that watch for whatever reason sometimes fail?

Whoever replicates this consitently, try to play with https://github.com/argoproj/argo-cd/blob/5b79c34c72300e6e2e6336051ce6992f6d54011c/controller/cache/cache.go#L50

**Comment by gn-viet-n** [comment]
I'm experiencing the same problem with latest argocd chart version "7.8.7"

**Comment by crenshaw-dev** [comment]
For folks waiting for this: we expect the fix to ship with 3.0.

**Comment by mike-code** [comment]
> For folks waiting for this: we expect the fix to ship with 3.0.

is the issue already known so that we can backport it to 2.x?

**Comment by zeppelinen** [comment]
> For folks waiting for this: we expect the fix to ship with 3.0.

Can you share a link to a specific piece of code?

**Comment by crenshaw-dev** [comment]
This is the fix: https://github.com/argoproj/gitops-engine/pull/646

It's too invasive to cherry-pick, so it has to wait for 3.0.

**Comment by dejanzele** [comment]
The issue appears when an Object in a Hook phase gets deleted externally (i.e. like how the Job controller deletes the Job object if TTL is set) and the gitops engine logic fails to notice that and gets stuck progressing.

The fix is to attach a finalizer for Hook objects and remove it after ArgoCD detects the hook phase terminated.

cc @mike-code

**Comment by David2011Hernandez** [comment]
> The issue appears when an Object in a Hook phase gets deleted externally (i.e. like how the Job controller deletes the Job object if TTL is set) and the gitops engine logic fails to notice that and gets stuck progressing.
> 
> The fix is to attach a finalizer for Hook objects and remove it after ArgoCD detects the hook phase terminated.
> 
> cc [@mike-code](https://github.com/mike-code)

Can you elaborate on this solution please. I am waiting for argocd 3.x version but in the meantime would be nice to have a workaround

**Comment by dejanzele** [comment]
@David2011Hernandez hook objects now have a finalizer which gets removed when the hook phase terminates.

**Comment by mhrcm** [comment]
still having this issue in 3.0

**Comment by younsl** [comment]
Is there a clear solution?

I'm facing sync issue from [ingress-nginx 1.12.3](https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v1.12.3) controlled by ArgoCD v2.13.4. The ingress-nginx-admission-create job remains in "Pending deletion" status indefinitely even after completion, and the ingress-nginx application also waits indefinitely in "Syncing" status.

ingress-nginx sync status in Argo UI:

<img width="1176" alt="Image" src="https://github.com/user-attachments/assets/52ccaf17-8a18-4dd3-9138-eda6593ffbe5" />

```bash
waiting for completion of hook batch/Job/ingress-nginx-admission-create
```

**Comment by cnd4** [comment]
@younsl as far as I understand, the solution is to upgrade to ArgoCD 3.

**Comment by Vacxe** [comment]
<img width="381" alt="Image" src="https://github.com/user-attachments/assets/83184583-b65f-4120-a0f2-dd05777bc0fd" />

Ver: v3.0.3+a14b012

**Comment by NPastorale** [comment]
I am seeing the same thing. After upgrading to v3 this is still getting stuck in at least some of the applications that have a helm chart as a source. I'm currently running v3.0.6+db93798

**Comment by dejanzele** [comment]
@NPastorale @Vacxe can you share the charts which are stuck (assuming you are using public charts)?

**Comment by NPastorale** [comment]
Sure. First and foremost  `ArgoCD`
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: argocd
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  destination:
    namespace: argocd
    server: https://kubernetes.default.svc
  project: default
  source:
    chart: argo-cd
    helm:
      valuesObject:
        global:
          domain: argocd.example.com
        configs:
          params:
            server.insecure: true
        server:
          ingress:
            enabled: true
            ingressClassName: internal-nginx
            tls: true
          ingressGrpc:
            enabled: true
            ingressClassName: internal-nginx
            tls: true
          certificate:
            enabled: "true"
            issuer:
              group: "cert-manager.io"
              kind: "ClusterIssuer"
              name: "letsencrypt-prod"
        extraObjects:
          - apiVersion: networking.k8s.io/v1
            kind: Ingress
            metadata:
              name: argocd-server-webhook-ingress
              annotations:
                cert-manager.io/cluster-issuer: "letsencrypt-prod"
                external-dns.alpha.kubernetes.io/exclude: "true"
            spec:
              ingressClassName: external-nginx
              tls:
                - hosts:
                    - argocd.example.com
                  secretName: argocd-ingress-webhook
              rules:
                - host: argocd.example.com
                  http:
                    paths:
                      - path: /api/webhook
                        pathType: Prefix
                        backend:
                          service:
                            name: argocd-server
                            port:
                              name: http
    repoURL: https://argoproj.github.io/argo-helm
    targetRevision: 8.1.0
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```
`Ingress-nginx` is another one that sees somewhat frequent updates and gets stuck
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: external-ingress
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  destination:
    namespace: external-ingress-nginx
    server: https://kubernetes.default.svc
  project: default
  source:
    chart: ingress-nginx
    helm:
      parameters:
        - name: controller.ingressClassResource.controllerValue
          value: "k8s.io/external-nginx"
        - name: controller.ingressClassResource.name
          value: "external-nginx"
        - name: controller.replicaCount
          value: "3"
        - name: controller.ingressClassResource.default
          value: "true"
        - name: controller.ingressClassByName
          value: "true"
        - name: controller.config.hsts-max-age
          value: "31536000"
        - name: controller.config.hsts-preload
          value: "true"
        - name: controller.config.allow-snippet-annotations
          value: "true"
        - name: controller.config.enable-real-ip
          value: "true"
        - name: controller.service.loadBalancerIP
          value: "10.10.20.56"
        - name: controller.service.externalTrafficPolicy
          value: "Local"
    repoURL: https://kubernetes.github.io/ingress-nginx
    targetRevision: 4.12.3
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```
Lastly `GitLab` is giving me the most issues, there are jobs that get created when issuing a `Sync` which get terminated almost instantly before being able to succeed (not sure if this is the same issue, but the symptom sure is)
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: gitlab
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  destination:
    namespace: gitlab
    server: https://kubernetes.default.svc
  project: default
  source:
    chart: gitlab
    repoURL: https://charts.gitlab.io/
    targetRevision: 9.0.2
    helm:
      valuesObject:
        upgradeCheck:
          enabled: false
        global:
          appConfig:
            defaultColorMode: 3
            defaultTheme: 2
            defaultSyntaxHighlightingTheme: 5
            terraformState:
              enabled: true
          edition: ee
          hosts:
            domain: example.com
            https: true
          ingress:
            configureCertmanager: false
            class: internal-nginx
            annotations:
              cert-manager.io/cluster-issuer: "letsencrypt-prod"
          kas:
            enabled: false
          pages:
            enabled: true
          psql:
            database: app
            username: app
            host: postgres-gitlab-rw
            password:
              secret: postgres-gitlab-app
              key: password
          minio:
            enabled: true
        minio:
          persistence:
            enabled: true
            storageClass: "nfs-csi-persistent"
            size: 100Gi
        redis:
          master:
            persistence:
              storageClass: "nfs-csi-persistent"
              size: 50Gi
        certmanager:
          install: false
        nginx-ingress:
          enabled: false
        postgresql:
          install: false
        prometheus:
          install: false
        gitlab-runner:
          install: true
          gitlabUrl: "https://gitlab.example.com"
          rbac:
            create: true
          checkInterval: 3
          runners:
            locked: false
            config: |
              [[runners]]
                url = "https://gitlab.example.com"
                clone_url = "https://gitlab.example.com"
                [runners.kubernetes]
                privileged = true
                image = "alpine:latest"
          podAnnotations:
            gitlab.com/prometheus_scrape: "true"
            gitlab.com/prometheus_port: 9252
        gitlab:
          gitaly:
            persistence:
              storageClass: "nfs-csi-persistent"
              size: 500Gi
          gitlab-pages:
            ingress:
              tls:
                secretName: pages-tls
          webservice:
            ingress:
              tls:
                secretName: gitlab-tls
        registry:
          ingress:
            tls:
              secretName: registry-tls
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```

Those are the ones I'm positive this can be reproduced with, although it is still somewhat random. Not every single time a Sync occurs they'll get stuck. I haven't been able to notice a pattern, if I just keep terminating the Sync process and doing it again, it eventually works.

**Comment by rverchere** [comment]
@younsl see https://github.com/kubernetes/ingress-nginx/pull/13483.

On my side, issues with nginx-ingress charts v4.11.7 & v4.12.2, where https://github.com/kubernetes/ingress-nginx/pull/13408 was added.

Rollback to v4.11.6 resolves temporarily my issue.

Using ArgoCD v2.14.11+8283115

**Comment by alekseikurepin** [comment]
> waiting for completion of hook batch/Job/ingress-nginx-admission-create

same think with "batch/Job/ingress-nginx-admission-create" and "batch/Job/ingress-nginx-admission-path" but for nginx-controller v4.11.7

**Comment by Kamaradeivanov** [comment]
@rverchere 👋 
I can confirm that upgrading ArgoCD to v3.0.6 fixes the issue on nginx-ingress v4.11.7.

**Comment by jkleinlercher** [comment]
For me this is not solved in argocd 3.0.6
@dejanzele ingress-nginx helm 4.12.3 and argocd 3.0.6

**Comment by htolic** [comment]
I find it working only with manual intervention. First terminate sync then click on ConfigMap (there is only one) and sync only that ConfigMap. It will sync and everything else will immediately start syncing.
Tho, I'm not sure why it works and would it help to put ConfigMap in some sync stage/wave to sync as early as possible

**Comment by mjtrangoni** [comment]
For me it works now with argocd 3.06 and ingress-nginx 4.12.4. 
NOTE: I synched the argocd-cm ConfigMap from the upstream one, also adding all `resource.customizations.ignoreResourceUpdates.*` configs there too.

**Comment by ElsevierAlex** [comment]
I’m experiencing the same issue:
`waiting for completion of hook batch/Job/ingress-nginx-admission-create`

Versions:
Ingress NGINX Helm chart: v4.12.4
Argo CD: v2.14.13

Terminating the sync and re-syncing does not resolve the issue. The hook job remains stuck in the "waiting for completion" state.

**Comment by SuperCoolAlan** [comment]
I found a solution for y'all! 🥳 

I am deploying `ingress-nginx` chart version 4.13.0. In my cluster's event logs, I saw that the job was being scheduled and no errors were reported. The pod & the job were then cleaned up promptly. I figure the Helm hook is poll-based and was not able to observe the job succeed before it was cleaned up by the cluster.

To address this, I had to patch the Helm chart's Job definition in `<chart-dir>/templates/admission-webhooks/job-patch/job-createSecret.yaml` at .spec.ttlSecondsAfterFinished from `0` to `10` to give the Helm controller enough time to see that the job had completed.

```yaml
# templates/admission-webhooks/job-patch/job-createSecret.yaml
...
spec:
  ttlSecondsAfterFinished: 10
```

**Comment by gihif** [comment]
> I found a solution for y'all! 🥳
> 
> I am deploying `ingress-nginx` chart version 4.13.0. In my cluster's event logs, I saw that the job was being scheduled and no errors were reported. The pod & the job were then cleaned up promptly. I figure the Helm hook is poll-based and was not able to observe the job succeed before it was cleaned up by the cluster.
> 
> To address this, I had to patch the Helm chart's Job definition in `<chart-dir>/templates/admission-webhooks/job-patch/job-createSecret.yaml` at .spec.ttlSecondsAfterFinished from `0` to `10` to give the Helm controller enough time to see that the job had completed.
> 
> # templates/admission-webhooks/job-patch/job-createSecret.yaml
> ...
> spec:
>   ttlSecondsAfterFinished: 10

Thanks, @SuperCoolAlan 
this solution works for me.

I patched the helm chart using a kustomize like this.
```yaml
helmCharts:
  - name: ingress-nginx
    releaseName: ingress-nginx
    repo: https://kubernetes.github.io/ingress-nginx
    version: 4.13.0
    namespace: ingress-nginx
    valuesFile: values.yaml

patches:
  - patch: |-
      - op: replace
        path: /spec/ttlSecondsAfterFinished
        value: 10
    target:
      kind: Job
      name: ingress-nginx-admission-create
  - patch: |-
      - op: replace
        path: /spec/ttlSecondsAfterFinished
        value: 10
    target:
      kind: Job
      name: ingress-nginx-admission-patch

```

**Comment by jurgenweber** [comment]
yeah, so in the end we should just make a PR against ingress nginx and make ttlSecondsAfterFinished configuration in the values.

**Comment by davoustp** [comment]
Looks like the ingress-nginx team refused such capability: look at https://github.com/kubernetes/ingress-nginx/pull/13483

**Comment by jurgenweber** [comment]
wow, that is wildly rude

**Comment by crenshaw-dev** [comment]
@jurgenweber let's avoid cross-org commentary on other open source project maintainers' tones or ways of engaging with their community. There are venues to discuss those things, but I don't think this thread is one of those venues. Those maintainers provide a huge service to a lot of people. While their particular decision to not support `ttlSecondsAfterFinished` may cause some short-term pain for Argo CD users, we should evaluate their decision in light of their more in-depth knowledge of what it takes to run their project well and in light of their project's overall contribution to the community.

I see folks have started brainstorming mitigations on the ingress-nginx thread. Mutating webhooks and Kustomize overlays may be good stop-gaps for many.

On this org, I'd like to understand more about folks' struggles to upgrade to Argo CD 3.0. It was intended to be a very simple upgrade, so if we've failed to meet that bar, let's start enumerating the problems and tackling them.

I also saw some mentions that maybe the finalizer solution hasn't completely solved the ingress-nginx use case. I'd like to get more details about that if possible. We've seen some initial problems with the finalizers and @agaudreault has spent some time trying to tidy those up. Hopefully we can improve the situation for everyone as quickly as possible.

**Comment by jurgenweber** [comment]
>On this org, I'd like to understand more about folks' struggles to upgrade to Argo CD 3.0. It was intended to be a very simple upgrade, so if we've failed to meet that bar, let's start enumerating the problems and tackling them.

For me, it is on the todo list but other things take prio right now.

**Comment by davoustp** [comment]
Hi @crenshaw-dev,
I concur with the added value of both projects to the community, no debate about how excellent those are, from where I stand.
As a fervent OSS proponent, all those rock!

> On this org, I'd like to understand more about folks' struggles to upgrade to Argo CD 3.0. It was intended to be a very simple upgrade, so if we've failed to meet that bar, let's start enumerating the problems and tackling them.

I suppose that upgrading argocd is not a big deal on its own (and btw, what a tremendous effort and details on the migration guide, this is awesome!).
Problem usually comes when dealing with daily dev/cloud/sec/ops tasks and priorities, upgrading such or such component is usually driven by needs and risks / benefits balance - no surprise here.
As such, the ingress-nginx component directly impacts public-facing services, and is also more exposed to sec risks and CVEs. And that's the case with many other components, and since argocd is the "pilot behind the wheel" of it all, it unfortunately (and wrongly) stands into the tail of the queue: even though the pilot is aging, people usually prefer changing the failing tires and engine rather than upgrading the driver ;-)

Technically, I agree that the finalizer route for argocd is the best one, as it brings a solid and consistent resource operations outcome monitoring to argocd.
So the whole purpose of controlling this (already but hardcoded) existing helm chart parameter in ingress-nginx was a very lightweight and straightforward way to deal with this situation when encountered for any of the reasons above - with zero impact on the ingress-nginx runtime itself.
Deploying a mutating webhook is a no-no for me (one more thing to deploy and maintain), and using a kustomize overlay may also prove difficult when no kustomize capability is available in the deployment process.

In any case, decision's been made, and therefore upgrading to argocd 3.x is the way forward!

Thanks again for the tremendous work.

**Comment by crenshaw-dev** [comment]
Kind of an aside, but a bit relevant: I've become more and more a fan of using Kustomize with a Helm chart inflator for basically every Helm chart. One never knows when they'll need to override something that isn't configurable in the chart. It's also possible that Argo CD could be more helpful by providing a generic Kustomize patch mechanism directly from the Application spec. But I tend to favor relying on the external tools.

**Comment by davoustp** [comment]
> I've become more and more a fan of using Kustomize with a Helm chart inflator for basically every Helm chart

@crenshaw-dev would you have a resource to share on what you implemented / use? Could prove useful beyond this very topic, indeed.

**Comment by crenshaw-dev** [comment]
Here's info about the [Helm chart inflator](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/) Kustomize feature. Personally, I'd use it basically any time I used Helm.

**Comment by loafoe** [comment]
@crenshaw-dev we use this pattern of kustomize -> helm extensively. We  combine it with an `envsubst` kustomize plugin that substitutes cluster/account specific placeholders (e.g. `${AWS_ACCOUNT_ID}` ${OIDC_ISSUER}`, etc) . 
That way we keep our kustomize + helm generic enough so we can apply them to different clusters without any changes.

**Comment by chrislin22** [comment]
> Kind of an aside, but a bit relevant: I've become more and more a fan of using Kustomize with a Helm chart inflator for basically every Helm chart. One never knows when they'll need to override something that isn't configurable in the chart. It's also possible that Argo CD could be more helpful by providing a generic Kustomize patch mechanism directly from the Application spec. But I tend to favor relying on the external tools.

we have been using argocd with `helmfile` `cmp`, which combines the power of `helm` and `kustomize`, which works out pretty well for us

**Comment by Boniker** [comment]
Hello everyone, +1

We are still experiencing this issue intermittently on ArgoCD **v3.2.6+65b0293** (jfyi: @crenshaw-dev @dejanzele)

Symptoms:
- Application _randomly_ gets stuck with: `waiting for completion of hook batch/Job/[job-name]`

<img width="311" height="87" alt="Image" src="https://github.com/user-attachments/assets/41ec8f1e-aa9b-490a-9417-639c312f7f24" />

- Job:
  - completes successfully in Kubernetes (confirmed via events)
  - disappears before ArgoCD observes completion
- This issue happens occasionally during automated syncs (Kargo promotions)

Job hook configuration:

```
metadata:
  annotations:
    argocd.argoproj.io/hook: PostSync
    argocd.argoproj.io/hook-delete-policy: HookSucceeded
    argocd.argoproj.io/sync-options: Force=true
spec:
  # No ttlSecondsAfterFinished set (not 0, it's not define at all)
```

Observations:
- Job execution time: ~3-5 seconds
- No deletion events in Kubernetes logs - job just disappears

As an idea: possible issue could be that it's being pruned by ArgoCD before status is observed

**Comment by afer4152** [comment]
> ConfigMap https://github.com/argoproj/argo-cd/issues/6880#issuecomment-3051403310

This works for me

**Comment by nvtkaszpir** [comment]
another option to workaround it is to starve cpu on the jobs:
```
controller:
  admissionWebhooks:
    createSecretJob:
      resources:
        requests:
          cpu: 1m
        limits:
          cpu: 1m
    patchWebhookJob:
      resources:
        requests:
          cpu: 1m
        limits:
          cpu: 1m
```

:D
